### PR TITLE
Whisker based report building structure with example

### PR DESCRIPTION
**Pull request by thongsav-usgs**
See the README.md file in the PR for the overall process of creating a report.

### CODE REVIEW

File: R/utils-render.R [hunk 1]
```diff
@@ -57,4 +70,185 @@ startRender <- function(data, author, reportName){
...
+#' 
+#' Peform the rendering of subfragments for a given report, should return a list of html fragments to 
+#' be available to inject into mustach templates
+#'
+#' @param reportData typed report json, the type should be the name of the report
+#' @return vector of rendered HTML fragments that will be included in the data available to the final report template. The fragments will be
+#' available to the templates as renderedFragments.html
+#' 
+#' @export
+renderCustomFragments <- function(reportData) UseMethod("renderCustomFragments")
```
Comment (jiwalker-usgs) [inline R/utils-render.R]: I showed you S3 objects, but this project already has S4 in place, it is probably best not to mix the two.

Comment (thongsav-usgs) [reply R/utils-render.R]: 👍 I'll make the conversion to S4.

File: R/02-templating-class.R [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#' 
+#' Peform the rendering of subfragments for a given report, should return a list of html fragments to 
+#' be available to inject into mustach templates
+#'
+#' @param reportData typed report json, the type should be the name of the report
+#' @return vector of rendered HTML fragments that will be included in the data available to the final report template. The fragments will be
+#' available to the templates as renderedFragments.html
+#' 
+#' @export
+renderCustomFragments <- function(reportData) 0
```
Comment (jiwalker-usgs) [inline R/02-templating-class.R:11]: Is this what they suggest for stubbing in the function?

File: R/utils-render.R [hunk 3]
```diff
@@ -27,16 +45,11 @@ cleanTempSpace <- function() {
...
 #' 
 startRender <- function(data, author, reportName){
-  data <- data 
-  
-  wd <- getwd()
-  tmp_folder_name <- paste0('tmp-', reportName)
+  warning("This rendering style is deprecated, convert to using whisker style and startTemplatedRender")
   
-  output_dir <- paste0(wd, "/", tmp_folder_name)
+  data <- data
```
Comment (jiwalker-usgs) [inline R/utils-render.R:50]: does this actually do anything?

**Review by jiwalker-usgs — COMMENTED** [review]
S4 is a bit goofy this way, which is partially why I prefer S3, but I think this looks alright.

**Review by zmoore-usgs — COMMENTED** [review]
Looks good to me, just a couple of comments:

1. Not sure if any of the functions added to utils-render are really testable, but if they are we should probably add tests for them.

2. Since the JSON parameter names for the header are now in a shared header we should probably go back and make sure that we have those names consistent in the report choreographers because I'm not sure if they are. Perhaps it would be good to add some sort of shared function for generating the report metadata or a shared list of the required metadata parameters for every report so that we don't have inconsistencies across the reports.

File: R/utils-render.R (hunk 3 above)
Comment (thongsav-usgs) [reply R/utils-render.R:50]: I think it puts data in the env so it's available to rmarkdown. I never quite wrapped my head around why but this is going away soon.

**Comment by thongsav-usgs** [comment]
@zmoore-usgs, so far the new functions are only tested via the integration test in test-renderReport, but I will look at adding unit tests.

The reportMetadata is definitely NOT consistent across reports, in the header you'll see there are template fragments for UV Hydro. We will do consolidation as we convert reports from rmarkdown to whisker.

File: R/02-templating-class.R (hunk 2 above)
Comment (thongsav-usgs) [reply R/02-templating-class.R:11]: Yeah, this was suggested as the base function. Without it the setMethod/setGeneric complains that there isn't a function yet defined.

**Pull request merged**